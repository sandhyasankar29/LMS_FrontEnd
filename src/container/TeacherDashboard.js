import { React, useState, useEffect } from 'react';
import axios from 'axios';
import Swal from 'sweetalert2'
import {
    Box,
    Button,
    Grommet,
    Text,
    Accordion, AccordionPanel,
    Header, Nav,
    Anchor, Form,
    FormField, TextInput,
    TextArea, Layer
} from 'grommet';
import { FormClose, Info, StatusGood, User, Search } from 'grommet-icons';
import { grommet } from 'grommet/themes';
import LogOut from './LogOut';
import './GlobalVariables'
const TeacherDashboard = () => {
    const urlTeacherLeave = 'http://'+{global,ip}.ip+':3001/teacherleaverecords/';
    const urlChairLeave = 'http://'+{global,ip}.ip+':3001/chairleaverecords/';
    const urlDecision = 'http://'+{global,ip}.ip+':3001/decisionteacher';
    const urlNotify = 'http://'+{global,ip}.ip+':3001/notifystudent';
    const Toast = Swal.mixin({
        toast: true,
        position: 'top-end',
        showConfirmButton: false,
        timer: 3000,
        timerProgressBar: true,
        didOpen: (toast) => {
            toast.addEventListener('mouseenter', Swal.stopTimer)
            toast.addEventListener('mouseleave', Swal.resumeTimer)
        }
    })
    const [count, changeCount] = useState(0)
    const [search, setSearch] = useState("");
    useEffect(() => {
        (localStorage.getItem("chairperson") !== 'Yes') ? axios.get(urlTeacherLeave + localStorage.getItem("tuserid"))
            .then(res => {
                if (res.data !== "No Data Available as of Now") {
                    setStudent(res.data)
                }
                else {
                    setStudent(null)
                }
            }).catch(error => {
                Toast.fire({
                    icon: "error",
                    title: 'Detail Fetch Failure'
                })
            }) : axios.get(urlChairLeave + localStorage.getItem("tuserid").substring(2, 5))
                .then(res => {
                    if (res.data !== "No Data Available as of Now") {
                        setStudent(res.data)
                    }
                    else {
                        setStudent(null)
                    }
                }).catch(error => {
                    Toast.fire({
                        icon: "error",
                        title: 'Detail Fetch Failure'
                    })
                })
    }, [count, Toast]);
    const decisionSubmit = (decision, suserid, dos, doe, reason) => {
        if (localStorage.getItem("chairperson") === 'Yes') {
            if (decision === "Verified") {
                decision = "Approved"
            }
        }
        axios.post(urlDecision, { status: decision, suserid: suserid, dos: dos })
            .then(res => {
                if (decision === "Approved" || decision === "Rejected") {
                    axios.post(urlNotify, { status: decision, suserid: suserid, dos: dos, doe: doe, reason: reason })
                        .then(response => {
                            Toast.fire({
                                icon: "success",
                                title: 'SMS Sent'
                            })
                        })
                        .catch(err => {
                            Toast.fire({
                                icon: "error",
                                title: 'Failed to send SMS'
                            })
                        })
                }
                else {
                    Toast.fire({
                        icon: "success",
                        title: 'Decision Registered'
                    })
                }
            }).catch(error => {
                Toast.fire({
                    icon: "error",
                    title: 'Detail Fetch Failure'
                })
            })
        changeCount(count + 1)
    }

    const [student, setStudent] = useState()
    const [open, setOpen] = useState(true);
    const onClose = () => setOpen(undefined);
    const items = [
        { label: 'Dashboard', href: '/teacherdashboard', value: 0 },
        { label: 'Password Change', href: '/teacherchangepassword', value: 1 },
        { label: 'Button', href: '#', value: 2 },
    ];
    return (
        <Grommet full theme={grommet} style={{ overflowY: 'scroll' }}>
            <Header background="dark-1" pad="small">
                <Box direction="row" align="center" gap="small">
                    <User />
                    <Anchor color="white" href="#">
                        Welcome {localStorage.getItem('firstname')}
                    </Anchor>
                    <TextInput icon={<Search />} placeholder="search ..." onChange={(event) => setSearch(event.target.value)} />
                </Box>
                <Nav direction="row">
                    {items.map(item => (
                        item.value !== 2 ? <Anchor href={item.href} label={item.label} key={item.label} /> : <LogOut route={'/teachersignin'} />
                    ))}
                </Nav>
            </Header>
            <div class="flex items-center justify-center pa4 bg-yellow">
                <Info />
                <span class="lh-title ml3">The following are the Student Leave Details, Please verify and take appropriate action</span>
            </div>
            <Box>
                {student ? student.filter(item => item.suserid.toLowerCase().includes(search.toLowerCase())).map((item) => (
                    <Accordion>
                        <AccordionPanel key={item} label={item.suserid} id="studentNo">
                            <Box align="center" pad="large" id="application">
                                <Form>
                                    <Box border gap="medium" pad="large" width="medium" id="leaveRequest">
                                        {localStorage.getItem('chairperson') === 'Yes' ? <FormField
                                            htmlFor="info-id"
                                            name="info-demo"
                                            label="Verified By"
                                        >
                                            <TextInput
                                                id="info-id"
                                                value={item.tuserid}
                                            />
                                        </FormField> : null}
                                        <FormField
                                            htmlFor="info-id"
                                            name="info-demo"
                                            label="Student ID"
                                        >
                                            <TextInput
                                                id="info-id"
                                                value={item.suserid}
                                            />
                                        </FormField>
                                        <FormField
                                            htmlFor="info-id"
                                            name="info-demo"
                                            label="Date of Start"
                                        >
                                            <TextInput
                                                id="dos"
                                                value={new Date(item.dos).toLocaleDateString("sq-AL", { year: 'numeric', day: '2-digit', month: '2-digit' })}
                                            />
                                        </FormField>
                                        <FormField
                                            htmlFor="info-id"
                                            name="info-demo"
                                            label="Date of End"
                                        >
                                            <TextInput
                                                id="doe"
                                                value={new Date(item.doe).toLocaleDateString("sq-AL", { year: 'numeric', day: '2-digit', month: '2-digit' })}
                                            />
                                        </FormField>
                                        <FormField
                                            htmlFor="info-id"
                                            name="info-demo"
                                            label="Reason"
                                        >
                                            <TextArea
                                                id="reason"
                                                value={item.reason}
                                            />
                                        </FormField>
                                        {item.cert !== 'NA' ? <FormField
                                            htmlFor="info-id2"
                                            name="info-demo2"
                                            label="Medical Certificate"
                                        >
                                            <TextArea
                                                value={item.cert}
                                            />
                                        </FormField> : null}

                                        <Box direction="row" justify="between">
                                            <Button id="approveBtn" label="Approve" style={{ backgroundColor: '#4BB543' }} onClick={() => decisionSubmit("Verified", item.suserid, new Date(item.dos).toLocaleDateString("sv-SE", { year: 'numeric', day: '2-digit', month: '2-digit' }), new Date(item.doe).toLocaleDateString("sv-SE", { year: 'numeric', day: '2-digit', month: '2-digit' }), item.reason)} />
                                            <Button id="rejectBtn" label="Reject" style={{ backgroundColor: '#FF6347' }} onClick={() => decisionSubmit("Rejected", item.suserid, new Date(item.dos).toLocaleDateString("sv-SE", { year: 'numeric', day: '2-digit', month: '2-digit' }), new Date(item.doe).toLocaleDateString("sv-SE", { year: 'numeric', day: '2-digit', month: '2-digit' }), item.reason)} />
                                        </Box>
                                    </Box>
                                </Form>
                            </Box>
                        </AccordionPanel>
                    </Accordion>
                )) : null}
            </Box>
            {
                open && <Layer
                    position="bottom"
                    modal={false}
                    margin={{ vertical: 'medium', horizontal: 'small' }}
                    onEsc={onClose}
                    responsive={false}
                    plain
                >
                    <Box
                        align="center"
                        direction="row"
                        gap="small"
                        justify="between"
                        round="medium"
                        elevation="medium"
                        pad={{ vertical: 'xsmall', horizontal: 'small' }}
                        background="status-ok"
                    >
                        <Box align="center" direction="row" gap="xsmall">
                            <StatusGood />
                            <Text>
                                Succcesfully Signed In To Teacher Portal
              </Text>
                        </Box>
                        <Button icon={<FormClose />} onClick={onClose} plain />
                    </Box>
                </Layer>
            }
        </Grommet>
    );
};

export default TeacherDashboard;