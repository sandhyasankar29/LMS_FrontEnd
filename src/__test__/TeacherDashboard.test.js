import React from 'react';
import ReactDOM from 'react-dom';
import TeacherDashboard from '../container/TeacherDashboard';
import { cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'

afterEach(cleanup)
it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<TeacherDashboard></TeacherDashboard>, div)
})

it("matches snapshot", () => {
    const tree = renderer.create(<TeacherDashboard></TeacherDashboard>).toJSON();
    expect(tree).toMatchSnapshot();
})
