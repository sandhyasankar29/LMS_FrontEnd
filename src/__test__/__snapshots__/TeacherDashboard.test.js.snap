// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="StyledGrommet-sc-19lkkz7-0 bWDPZI"
  style={
    Object {
      "overflowY": "scroll",
    }
  }
>
  <header
    className="StyledBox-sc-13pk1d4-0 jwhdBG"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 dfYOnT"
    >
      <svg
        aria-label="User"
        className="StyledIcon-ofa7kd-0 dcacLZ"
        viewBox="0 0 24 24"
      >
        <path
          d="M8,24 L8,19 M16,24 L16,19 M3,24 L3,19 C3,14.0294373 7.02943725,11 12,11 C16.9705627,11 21,14.0294373 21,19 L21,24 M12,11 C14.7614237,11 17,8.76142375 17,6 C17,3.23857625 14.7614237,1 12,1 C9.23857625,1 7,3.23857625 7,6 C7,8.76142375 9.23857625,11 12,11 Z"
          fill="none"
          stroke="#000"
          strokeWidth="2"
        />
      </svg>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 eBVJAG"
      />
      <a
        className="StyledAnchor-sc-1rp7lwl-0 ePWGdF"
        href="#"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        Welcome 
      </a>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 eBVJAG"
      />
      <div
        className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 dpaJQp"
      >
        <div
          className="StyledTextInput__StyledIcon-sc-1x30a0s-3 kdffHW"
        >
          <svg
            aria-label="Search"
            className="StyledIcon-ofa7kd-0 dcacLZ"
            viewBox="0 0 24 24"
          >
            <path
              d="M15,15 L22,22 L15,15 Z M9.5,17 C13.6421356,17 17,13.6421356 17,9.5 C17,5.35786438 13.6421356,2 9.5,2 C5.35786438,2 2,5.35786438 2,9.5 C2,13.6421356 5.35786438,17 9.5,17 Z"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </svg>
        </div>
        <input
          autoComplete="off"
          className="StyledTextInput-sc-1x30a0s-0 eQRxlb"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="search ..."
        />
      </div>
    </div>
    <div
      className="StyledBox__StyledBoxGap-sc-13pk1d4-1 ffHQai"
    />
    <nav
      className="StyledBox-sc-13pk1d4-0 glIhNO"
    >
      <a
        className="StyledAnchor-sc-1rp7lwl-0 eaoGTv"
        href="/teacherdashboard"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        Dashboard
      </a>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 ffHQai"
      />
      <a
        className="StyledAnchor-sc-1rp7lwl-0 eaoGTv"
        href="/teacherchangepassword"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        Password Change
      </a>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 ffHQai"
      />
      <button
        className="StyledButton-sc-323bzc-0 bZMlsb"
        id="logoutBtn"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        type="button"
      >
        <svg
          aria-label="Logout"
          className="StyledIcon-ofa7kd-0 eBBIEF"
          viewBox="0 0 24 24"
        >
          <path
            d="M13,9 L13,2 L1,2 L1,22 L13,22 L13,15 M22,12 L5,12 M17,7 L22,12 L17,17"
            fill="none"
            stroke="#231F20"
            strokeWidth="2"
          />
        </svg>
      </button>
    </nav>
  </header>
  <div
    class="flex items-center justify-center pa4 bg-yellow"
  >
    <svg
      aria-label="Info"
      className="StyledIcon-ofa7kd-0 dEeaTZ"
      viewBox="0 0 24 24"
    >
      <path
        d="M15,17.0002864 C15,14.0000003 19,12.0005727 19,8.00028636 C19,4.00000002 16,1.00028636 12,1.00028636 C8,1.00028636 5,4.00000002 5,8.00028636 C5,12.0005727 9,14.0000003 9,17.0002864 C9,20.0005725 9,20 9,20 C9,22.0000003 10,22.9999997 12,23 C14,23.0000003 15,22.0000003 15,20 C15,20 15,20.0005725 15,17.0002864 Z M9,18 L15,18"
        fill="none"
        stroke="#000"
        strokeWidth="2"
      />
    </svg>
    <span
      class="lh-title ml3"
    >
      The following are the Student Leave Details, Please verify and take appropriate action
    </span>
  </div>
  <div
    className="StyledBox-sc-13pk1d4-0 dYebPD"
  />
</div>
`;
